DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/restartable_streaming_kernel/README.md

```diff
@@ -0,0 +1,282 @@
+
+```c++
+[[intel::initiation_interval(1)]]  // NO-FORMAT: Attribute
+while (keep_going) {
+  // Use non-blocking operations to ensure that the kernel can check all its
+  // pipe interfaces every clock cycle, even if one or more data interfaces
+  // are stalling (asserting valid = 0) or back-pressuring (asserting ready
+  // = 0).
+  bool did_write = false;
+  PipePayloadType beat = <...>;
+  OutputPipe::write(beat, did_write);
+
+  // Only adjust the state of the kernel if the pipe write succeeded.
+  // This is logically equivalent to blocking.
+  if (did_write) {
+    counter++;
+  }
+
+  // Use non-blocking operations to ensure that the kernel can check all its
+  // pipe interfaces every clock cycle.
+  bool did_read_keep_going = false;
+  bool stop_result = StopPipe::read(did_read_keep_going);
+  if (did_read_keep_going) {
+    keep_going = !stop_result;
+  }
+}
+```
+
+In this sample, `StopPipe` has been assigned the `protocol::avalon_mm_uses_ready` property so it is mapped to the kernel's control/status register (CSR) instead of a streaming interface. Mapping to the CSR allows this kernel to be managed by a memory-mapped host (such as a Nios® V softcore processor), but mapping to a streaming interface is convenient if this kernel were to be managed by another SYCL kernel. For details about the `protocol::avalon_mm_uses_ready` property, see the [CSR Pipes](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/component_interfaces_comparison/csr-pipes) sub-sample within the [Component Interfaces Comparison](/DirectProgramming/C++SYCL_FPGA/Tutorials/Features/hls_flow_interfaces/component_interfaces_comparison) code sample.
+
+![](assets/stopcsr.png)
+
+The testbench in `main.cpp` exercises the kernel in the following steps:
+
+1. Initialize the counter kernel with an initial value of 7.
+2. Read a sequence of 256 outputs from the kernel, which should be a monotonically growing sequence starting at 7.
+3. Read 256 more outputs from the kernel, which should be a monotonically growing sequence starting at 263.
+4. Stop the kernel.
+5. Initialize the kernel with a new initialization value of 77.
+6. Read 256 more outputs from the kernel, which should be a monotonically growing sequence starting at 77.
+
+### Packets
+
+This design uses the Avalon `start_of_packet` signal to indicate when the a new set of values is being written to `OutputPipe`. The `start_of_packet` sideband signal is not *generally* necessary for implementing a restartable streaming kernel, but it is used in this design to compensate for the decoupled way that the `RestartableCounter` kernel executes with respect to the host code. Since the host code does not tell `RestartableCounter` how many values to write to `OutputPipe`, the `RestartableCounter` will continue to write to `OutputPipe` until either
```

Comment:
This paragraph feels incomplete given the lack of punctuation (I know, minor issue). Can you just add a colon to it?

Reply:
done